pkg/piecestore/rpc/server/store.go

```diff
@@ -44,19 +44,22 @@ func (s *Server) Store(reqStream pb.PieceStoreRoutes_StoreServer) (err error) {
 		return StoreError.New("Piece ID not specified")
 	}
 
-	// If we put in the database first then that checks if the data already exists
-	if err = s.DB.AddTTLToDB(pd.GetId(), pd.GetExpirationUnixSec()); err != nil {
-		return StoreError.New("Failed to write expiration data to database")
+	exists, err := s.DB.DoesTTLExist(pd.GetId())
```

Comment:
There's a logical race here. With multiple concurrent requests two Exists checks can succeed and hence double stored.

Reply:
How about changing to the file system?